kimchi/src/tests/xor.rs

```diff
@@ -31,58 +59,145 @@ fn create_test_constraint_system(bits: usize) -> ConstraintSystem<Fp> {
     ConstraintSystem::create(gates).build().unwrap()
 }
 
-fn test_xor(in1: u128, in2: u128, bits: usize) -> [Vec<PallasField>; COLUMNS] {
-    let cs = create_test_constraint_system(bits);
-    let witness = xor::create(in1, in2, bits);
-    for row in 0..xor::num_xors(bits) + 1 {
-        assert_eq!(
-            cs.gates[row].verify_witness::<Vesta>(row, &witness, &cs, &witness[0][0..cs.public]),
-            Ok(())
-        );
+// Generates a random field element of up to a given number of bits
+pub(crate) fn random_field<G: KimchiCurve>(bits: usize, rng: &mut StdRng) -> G::ScalarField {
+    G::ScalarField::from_biguint(
+        rng.gen_biguint_range(&BigUint::from(0u8), &BigUint::from(2u8).pow(bits as u32)),
+    )
+    .unwrap()
+}
+
+// Initialize a random input
+pub(crate) fn initialize<G: KimchiCurve>(
+    input: Option<G::ScalarField>,
+    bits: Option<usize>,
+    rng: &mut StdRng,
+) -> G::ScalarField {
+    if let Some(inp) = input {
+        inp
+    } else {
+        assert!(bits.is_some());
+        let bits = bits.unwrap();
+        random_field::<G>(bits, rng)
     }
-    witness
 }
 
-// Returns a given crumb of 4 bits
-fn xor_crumb(word: u128, crumb: usize) -> u128 {
-    (word >> (4 * crumb)) % 2u128.pow(4)
+// Returns the all ones BigUint of bits length
+pub(crate) fn all_ones<G: KimchiCurve>(bits: usize) -> G::ScalarField {
+    G::ScalarField::from(2u128).pow(&[bits as u64]) - G::ScalarField::one()
 }
 
-// Manually checks the XOR of each crumb in the witness
-fn check_xor(witness: &[Vec<PallasField>; COLUMNS], bits: usize, input1: u128, input2: u128) {
-    for xor in 0..xor::num_xors(bits) {
+// Returns a given nybble of 4 bits
+pub(crate) fn xor_nybble(word: BigUint, nybble: usize) -> BigUint {
+    (word >> (4 * nybble)) % 2u128.pow(4)
+}
+
+// Manually checks the XOR of each nybble in the witness
+pub(crate) fn check_xor<G: KimchiCurve>(
+    witness: &[Vec<G::ScalarField>; COLUMNS],
+    bits: usize,
+    input1: G::ScalarField,
+    input2: G::ScalarField,
+    not: bool,
+) {
+    let input1 = input1.to_biguint();
+    let input2 = input2.to_biguint();
+    let ini_row = if not == XOR { 0 } else { 1 };
+    for x in 0..xor::num_xors(bits) {
         let in1 = (0..4)
-            .map(|i| xor_crumb(input1, i + 4 * xor))
-            .collect::<Vec<u128>>();
+            .map(|i| xor_nybble(input1.clone(), i + 4 * x))
+            .collect::<Vec<BigUint>>();
         let in2 = (0..4)
-            .map(|i| xor_crumb(input2, i + 4 * xor))
-            .collect::<Vec<u128>>();
-        for crumb in 0..4 {
+            .map(|i| xor_nybble(input2.clone(), i + 4 * x))
+            .collect::<Vec<BigUint>>();
+        for nybble in 0..4 {
             assert_eq!(
-                witness[11 + crumb][xor],
-                PallasField::from(in1[crumb] ^ in2[crumb])
+                witness[11 + nybble][x + ini_row],
+                G::ScalarField::from(BigUint::bitxor(&in1[nybble], &in2[nybble]))
```

Comment:
I'm confused about what `BigUint::bitxor` is doing because the output should be a `BigUint` while `ScalarField::from` should not expect a `BigUint` unless it was `ScalarField::from_biguint`

Reply:
it just returns the bitwise xor of two biguint inputs, and then i transform it to field format so i can compare with what's in the witness, but I will write it more readable